cid.go

```diff
@@ -132,36 +132,40 @@ var CodecToStr = map[uint64]string{
 // They exist to allow IPFS to work with Cids while keeping
 // compatibility with the plain-multihash format used used in IPFS.
 // NewCidV1 should be used preferentially.
-func NewCidV0(mhash mh.Multihash) *Cid {
-	return &Cid{
-		version: 0,
-		codec:   DagProtobuf,
-		hash:    mhash,
-	}
+func NewCidV0(mhash mh.Multihash) Cid {
+	return Cid{string(mhash)}
 }
 
 // NewCidV1 returns a new Cid using the given multicodec-packed
 // content type.
-func NewCidV1(codecType uint64, mhash mh.Multihash) *Cid {
-	return &Cid{
-		version: 1,
-		codec:   codecType,
-		hash:    mhash,
+func NewCidV1(codecType uint64, mhash mh.Multihash) Cid {
+	hashlen := len(mhash)
+	// two 8 bytes (max) numbers plus hash
+	buf := make([]byte, 2*binary.MaxVarintLen64+hashlen)
+	n := binary.PutUvarint(buf, 1)
+	n += binary.PutUvarint(buf[n:], codecType)
+	cn := copy(buf[n:], mhash)
+	if cn != hashlen {
+		panic("copy hash length is inconsistent")
 	}
+
+	return Cid{string(buf[:n+hashlen])}
 }
 
 // Cid represents a self-describing content adressed
 // identifier. It is formed by a Version, a Codec (which indicates
 // a multicodec-packed content type) and a Multihash.
-type Cid struct {
-	version uint64
-	codec   uint64
-	hash    mh.Multihash
-}
+// Byte layout: [version, codec, multihash]
+// - version uvarint
+// - codec uvarint
+// - hash mh.Multihash
+type Cid struct{ str string }
+
+var Nil = Cid{}
```

Comment:
How about calling this `Zero`?

"Zero" would more closely match https://golang.org/ref/spec#The_zero_value

Reply:
`Nil` is a more intuitive name to me.  `Zero` implies a numeric value to me.

Reply:
You are correct in that go would call this the "Zero" value. However, I really don't like it.

Arguments against Zero:

* To me, "Zero" implies a number.
* `Zero` could mean the CID for the number 0 (`z7xsiH4JqWQudech83`). I expect that users will define "constant" CIDs all the time although this one might be a stretch.
* We've historically used `nil` for "empty CIDs".
* We have a CIDv0 ("version zero").
* `Zero` "looks funny" (entirely subjective) in the return position.

```
func Foo() (Cid, error) {
    return Nil, fmt.Errorf("bad stuff")
}
func Bar() (Cid, error) {
    return Zero, fmt.Errorf("bad stuff")
}
```

Reply:
Some other options just to toss out:

- `cid.Invalid` -- this shows up in a lot of other idiomatic Go code... but mostly in enum situations, so maybe less appropriate here.
- `cid.None` -- avoids all the numeric connotations, less "looks funny" in returns I think?

Reply:
I still like `Nil` better...

Reply:
I could go for `Undef` :+1:

Reply:
:-1: on undef and invalid. The value *isn't* undefined and isn't the *only* invalid value. I'd prefer `None` (although I still prefer `Nil`).

Reply:
Actually, seeing `cid.Defined()`, I'm onboard with this (if we *return* `Undef` and *check* with `cid.Defined()`).

Reply:
Okay I am going with `Undef`.